firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/COBS.tpp

```diff
@@ -57,39 +63,42 @@ size_t encode_cobs(const uint8_t *buffer, size_t size, uint8_t *encoded_buffer)
 
   encoded_buffer[code_index] = code;
 
-  return write_index;
+  return IndexStatus::ok;
 }
 
-size_t decode_cobs(const uint8_t *encoded_buffer, size_t size, uint8_t *decoded_buffer) {
-  if (size == 0) {
-    return 0;
+template <size_t input_size, size_t output_size>
+IndexStatus decode_cobs(
+    const Util::ByteVector<input_size> &encoded_buffer,
+    Util::ByteVector<output_size> &decoded_buffer) {
+  if (encoded_buffer.empty()) {
+    return IndexStatus::out_of_bounds;
   }
 
   size_t read_index = 0;
-  size_t write_index = 0;
 
-  while (read_index < size) {
+  while (read_index < encoded_buffer.size()) {
     uint8_t code = encoded_buffer[read_index];
 
-    if (read_index + code > size && code != 1) {
-      return 0;
+    if (read_index + code > encoded_buffer.size() && code != 1) {
+      return IndexStatus::out_of_bounds;
     }
 
     read_index++;
 
     for (uint8_t i = 1; i < code; i++) {
-      decoded_buffer[write_index++] = encoded_buffer[read_index++];
+      uint8_t byte = encoded_buffer[read_index++];
+      decoded_buffer.push_back(byte);
     }
 
-    if (code != max_block_size + 1 && read_index != size) {
-      decoded_buffer[write_index++] = '\0';
+    if (code != max_block_size + 1 && read_index != encoded_buffer.size()) {
+      decoded_buffer.push_back(0x00);
```

Comment:
The deleted line incremented `write_index` after adding a null byte to `decoded_buffer`, do we not need to do the same thing in the new version?